### DISCUSSION TITLE
Fixing functional form

### DISCUSSION THREAD

**Discussion by kyoungyi** [MilesCranmer/PySR#291] [Q&A]
Hi Miles, thank you again for your swift reply and thorough answers about #273 .
I am still stuck with some problems and having a hard time figuring them out. Your help would be very appreciated again.

1. Poor predictive skills:
Despite using the default or recommended parameters you provided earlier (#273 #242 ), I still get poor R^2 values (e.g., around 0.1) and low Pearson correlations between the predicted and true response values. I expected the R^2 to be similar to or higher than the random forest model's R^2 (R^2 ~ 0.7) for the same training dataset. I am unsure what I am doing wrong here. Or could running the model for a longer time be a matter (https://github.com/MilesCranmer/PySR/discussions/275 )? You mentioned running it for a week but I noticed minor changes in the equations after a certain point (maybe after 70-80% progress). Do you have any suggestions for further tuning or improving the search?

2. Features selected by pysr:
I let all my features (~15 features) into the model as you suggested, but only a couple of variables that I thought were less important (based on science/references/RF model) repeatedly showed up in the final equations. For example, x1 and x2 dominate the equations out of 15 features: x1x1 +x1/x2 +sin(x1) + 0.02212x2. Any thoughts on this? How could I set the features show up one or two times with more various features? Any examples of codes here? I know you mentioned about customized loss function but I am unsure how to constrain it.

3. Formula prior:
I also need to fix a symbolic form of an expression with input variables (https://github.com/MilesCranmer/PySR/issues/285 ). For example, I wanted to have a form like f1(x1) + f2(x2,x3,x4,x5,x6,x7,x8,x9..) = y. In codes, I just changed (2,3) and (1,4) to the corresponding index number of my variables: e.g., "return !t.constant && t.feature in (1) and (2,3,4...9)". However, f2 has an x1 variable too and selects only a couple of variables. I could set the complexity of constants large enough so that only features can be included but I am still curious about what problems are here (maybe the same problem from question 2). Also, is there a way to control the complexity of f1 and f2 separately?

Any further input or recommendations would be very appreciated!

**Comment by MilesCranmer** [comment]
I will respond later, but just a quick note for 3 - would you be up for giving comments on this issue: https://github.com/MilesCranmer/SymbolicRegression.jl/issues/193? This proposes a more general way of imposing functional priors, and simplifies the process compared to writing a manual evaluation function. Since you are using the functional prior, it would be useful to hear what you think about that proposal!

---

Also, if it's okay with you, it might be helpful if you share your full search script + dataset (or a small sample of the dataset).

**Reply by kyoungyi** [reply]
Hi Miles, thanks for the head up. Yes, I will check the data confidentiality and come back to you. But if not able to be shared, I will share the full search script and the true vs. predicted value plots? FYI, I am not familiar with Julia, so I am not sure how much I can be helpful for that proposal 193. But thanks for asking and I will leave my thoughts after I think about it!

**Reply by kyoungyi** [reply]
Hi Miles, sorry for the delay in the reply. I had an extended exam for the last month so was not able to work on this one.

I do not think I can share the dataset yet so I am posting the codes here. Please feel free to make any suggestions.

To explain a little about the data, the response variable is the residuals after regressing on the state-year interaction variable (state is categorical and year is numerical here. Including them as the encoded variable in pysr does not make sense in my case as they do not include all state variables). 
As you recommended previously, I included all predictors regardless of the important feature outputs from the random forest model. I have 15 predictors and would like to separate "month" variable from the other 14 variables. Therefore, I am imposing functional priors as follows:
```
objective = """
function my_custom_objective(tree, dataset::Dataset{T}, options) where {T<:Real}
    # Require root node to be binary, so we can split it,
    # otherwise return a large loss:
    tree.degree != 2 && return T(10000)

    f1 = tree.l
    f2 = tree.r

    # Evaluate f1:
    f1_value, flag = eval_tree_array(f1, dataset.X, options)
    !flag && return T(10000)

    # Evaluate f2:
    f2_value, r_flag = eval_tree_array(f2, dataset.X, options)
    !r_flag && return T(10000)

    # Impose functional form:
    prediction = f1_value .+ f2_value

    # See if x2 or x3 in an expression:
    function contains_x2_x3(t)
        if t.degree == 0
            return !t.constant && t.feature in (15)
        elseif t.degree == 1
            return contains_x2_x3(t.l)
        else
            return contains_x2_x3(t.l) || contains_x2_x3(t.r)
        end
    end

    # See if x1 or x4 in an expression:
    function contains_x1_x4(t)
        if t.degree == 0
            return !t.constant && t.feature in (1,2,3,4,5,6,7,8,9,10,11,12,13,14)
        elseif t.degree == 1
            return contains_x1_x4(t.l)
        else
            return contains_x1_x4(t.l) || contains_x1_x4(t.r)
        end
    end

    f1_violating = contains_x2_x3(f1)
    f2_violating = contains_x1_x4(f2)

    regularization = T(1000) * f1_violating + T(100) * f2_violating

    prediction_loss = sum((prediction .- dataset.y) .^ 2) / dataset.n  ## mean squared error
    return prediction_loss + regularization
end
"""
```
```
model = PySRRegressor(procs=16,  maxsize=60, maxdepth =7,
                      populations = 48,
                      population_size = 65, 
                      niterations = 2000, # I know it is usually set to a large number but I am running from jupyter notebook which does not allow me to stop. so I set it to 2000 or less.
                      ncyclesperiteration = 2000,
                      weight_optimize = 0.001,
                      binary_operators=["+", "*", "-", "/","^"],
                     unary_operators=[
        "sin",
        "log",
    ], nested_constraints={
        "sin": {"sin":0,"log":0},#exp":0},
        "pow":{"pow":3,"sin":0,"div":0,"log":0},
        "log": {"log":0,"div":1,"sin":0,"pow":3},
        "div": {"div":0},
    }, 
    complexity_of_constants = 60, #testing not to include any constants at the moment since it was replacing month variable.
    model_selection ='accuracy', #chose accuracy rather than best, to get a better R^2, r, or AIC
    constraints={'pow': (-1, 3), 'mult': (8, 8), 'div':(8,6)}, 
    parsimony=0.05, #higher loss  
    adaptive_parsimony_scaling = 1000,
    precision=64,
    full_objective = objective)
```

I was using training data which was randomly distributed from the original dataset to (8158, 16), but now I am putting all data into the model to see if I get a better output. However, it takes a long time.. For example, from my local machine, getting 5 complexities took a day. I am also double-checking my data but would like to figure out if I am correctly configuring or not. I am considering different options including standardizing data #317. 

Will greatly appreciate your input/suggestions!

**Reply by MilesCranmer** [reply]
Hm, it might be because of the closure. Maybe try this?

```py
objective = """
# See if x2 or x3 in an expression:
function contains(t, features)
    if t.degree == 0
        return !t.constant && t.feature in features
    elseif t.degree == 1
        return contains(t.l, features)
    else
        return contains(t.l, features) || contains(t.r, features)
    end
end

function my_custom_objective(tree, dataset::Dataset{T}, options) where {T<:Real}
    # Require root node to be binary, so we can split it,
    # otherwise return a large loss:
    tree.degree != 2 && return T(10000)

    f1 = tree.l
    f2 = tree.r

    # Evaluate f1:
    f1_value, flag = eval_tree_array(f1, dataset.X, options)
    !flag && return T(10000)

    # Evaluate f2:
    f2_value, r_flag = eval_tree_array(f2, dataset.X, options)
    !r_flag && return T(10000)

    # Impose functional form:
    prediction = f1_value
    @. prediction = prediction + f2_value

    f1_violating = contains(f1, Tuple(1:14))
    f2_violating = contains(f2, (15,))

    regularization = 1000 * f1_violating + 100 * f2_violating

    loss(i) = (prediction[i] - dataset.y[i])^2  ## mean squared error
    prediction_loss = sum(loss, axes(prediction, 1)) / dataset.n
    return prediction_loss + T(regularization)
end

# This entire string is passed to `jl.eval()` in PyCall,
# so we can define functions (and import libraries) inside this, so long as the
# last thing evaluated is the function we want to pass, because
# jl.eval() returns the last thing evaluated.

my_custom_objective
"""
```

Also since you are ruling out constants with `complexity_of_constants=60`, I would set `weight_optimize=0.0` and `should_optimize_constants=False` and `weight_mutate_constant=0.0`.

**Reply by kyoungyi** [reply]
1. Fixed functional forms:
It was due to the closure. However, including constants in the search still makes it hard to follow the prior formula. For example, 1) in f2, a combination of constants replaced the predictor "month"; and 2) I had "month - month" in the expression which results in zero. I believe that constraining the complexities of constants and operators separately for multiple functions in the prior formula would greatly help in my case. However, I suppose I will have to wait for this implementation   https://github.com/MilesCranmer/SymbolicRegression.jl/issues/193. I am also curious about how @LP2006 is addressing this challenge for #292
 
2. Contraining constants and operators:
Previously, I excluded constants for testing purposes, because they were overpowering the expressions. Regardless of it, my search script still provides poor R^2 values, even for the training data, with most predictions being zeros. I also examined the outputs by plotting partial dependence plots, but they do not make sense to me. I am wondering if I am constraining the search too strongly or if I need to adjust other default parameters. Please feel free to make any comments on my search script!
```
PySRRegressor(procs=16,  maxsize=60, maxdepth =7,
                      populations = 48,
                      population_size = 65, 
                      niterations = 1000,
                      ncyclesperiteration = 1000,
                      ##weight_optimize = 0.001,
                      binary_operators=["+", "*", "-", "/"],
                     unary_operators=["sin", "tan", "square","cube"], 
nested_constraints={
        "sin": {"sin":0,"tan":0}, "tan":{"sin":0,"tan":0}, #sin and tan do not tend to be included in final expressions
        "square":{"square":0,"cube":0,"sin":0,"tan":0},
        "cube":{"square":0,"cube":0,"sin":0,"tan":0},
        "div": {"div":0}
        }, 
    complexity_of_constants = 3,
    model_selection ='accuracy', #chose accuracy rather than best, to get a better R^2, r, or AIC
    constraints={'square':1,'cube':1}, 
    parsimony=0.01,  
    adaptive_parsimony_scaling = 100,
    precision=64)
```

3. full_objective:
Is it possible to force the expression to have no global intercept? Basically, expressions can have operators with coefficients but not a global intercept. 

4. Slow search:
I found that the slower search came from higher adaptive_parsimony_scaling. Reducing this scaling helped me to get more expressions at higher complexities within a relatively short time. However, I was curious if this implies that the search is specializing too early in the wrong functional form (?) as mentioned in your paper. 

Apologies for adding more questions again! I really want to sort this out with pysr. 
Please feel free to correct my understanding or give any input on this matter.
